docs: Revamp info on how to add a new hub (#498)
2i2c-org/pilot-hubs: docs/howto/operate/add-hub.md

# Add a new hub

> ## Step 1: Decide where the hub will live

Author: I incorporated this, but pointing to pilot-hubs still. https://github.com/2i2c-org/team-compass/issues/151 talks about the support/ repository.

> ## Step 1: Decide where the hub will live

Reviewer: I think that is OK for now, @choldgraf?

Hubs need to exist on a kubernetes cluster, so we will need to

> decide which cluster it should live in.

Reviewer 2: ```suggestion
decide which cluster it should live in.
As a general rule, there is **one cluster per billing account**.
```

Author: Incorporated!

1. For standalone hubs where we are being paid or the client has cloud
   credits, we will need to create a cluster specifically for this hub
   before deploying it.
2. For hubs we are running for free, they will go into the `pilot-hubs`
   cluster.
3. Hubs run with the support of [cloudbank](https://www.cloudbank.org/),
   will go in the `cloudbank` cluster.

> 

Reviewer: I like this categorization!

Reviewer 3: +1!

Reviewer 2: I wonder if, rather than hard-coding "cloudbank" here, we should keep a list of "special" clusters like the CloudBank one (basically any cluster that has more than a 1-1 mapping onto customers) and then reference that table?

I seem to remember us having an issue about having such a "source of truth" for clusters, but maybe this was just for hubs?

Reviewer: It was for hubs, indeed (IIRC).
I am +1 about having something more dynamic here for "special" clusters...

Reviewer: Although maybe we can add that later... when we have more than one "special" cluster?

Each cluster has configuration + list of hubs it supports under
`config/hubs/<cluster-name>.cluster.yaml`.

## Step 2: Decide the template used for the hub

> Based on the kinda usage expected of the hub select a [hub template](../../topic/hub-templates.md)

Reviewer 2: ```suggestion
Hub templates map on to different major use-cases that we aim to enable.
Ask the hub user(s) about their needs and expected usage, and use their answer to select a [hub template](../../topic/hub-templates.md).
```

Reviewer 2: Additionally, perhaps we can document a basic list of the current user images and when each should be used? That feels like it could be another PR though...

to be used.

> 

Reviewer: Step 3 is missing here 😛

## Step 4: Decide the authentication provider to be used

> In consultation with the users, decide

Reviewer 2: On this page, perhaps we should define a **Community Champion** role for a hub (or maybe we should re-name this "user representative"?) and then suggest that we ask this person questions throughout the process, rather than to use a generic "user" identity like here? I'd like us to move away from the possibility that multiple users on a hub might be needed to answer questions.

Reviewer: > I'd like us to move away from the possibility that multiple users on a hub might be needed to answer questions.

💯

Reviewer: @choldgraf, until we have the Community champion role well defined, I think this is OK for now. Let us know if you disagree.

[which authentication provider](https://pilot.2i2c.org/en/latest/admin/configuration/login.html#authentication)

> the hub should use. While this can be changed later, it's a messy

Reviewer: Maybe convert the last sentence into a note?

process - so we should try to get this right the first time.

## Step 5: Decide the user image for the hub

The default user image is present in this repository (`images/user`),
and is geared towards simple datascience classes - with both R and
Python. Determine if this is acceptable - if not, you need to
make a [custom image](../configure/update-env.md) for them.

> 

Reviewer: Maybe a note about the upcoming process to self-serve the custom image?

Reviewer 2: yeah a TODO: here sounds good too (you can add comments in MyST markdown with %, so we can use that to add TODO similar to what we'd do in Python)

## Step 6: Add hub config

Add an entry for the hub in `config/hubs/<cluster-name>.cluster.yaml` for this hub.
The following docs might be helpful:

1. [Configuration Structure](../../topic/config.md)
2. [JSON Schema for `.cluster.yaml` file](https://github.com/2i2c-org/pilot-hubs/blob/master/config/hubs/schema.yaml)
3. [Zero to JupyterHub on k8s Docs](https://zero-to-jupyterhub.readthedocs.io/en/latest/), since ultimately
   that is what we are configuring.

## Step 7: Deploy manually and test

> [Deploy the hub manually](./manual-deploy.md) and test to make sure it works

Reviewer: We have automated deployment for the pilots and cloudbank hubs... are we encouraging here to do the manual deployment instead? Or we should include also the automatic deployment and then explain the testing steps indicated below?

Reviewer 2: Or more generally, it'd be useful to have some guide for when we should try a manual deploy vs. just auto-deploying vs. config.yaml. The manual deploy step I suspect will be a big amount of extra work per hub (and will mean that the only people that can add new hubs will also need to have the skills to manually deploy a hub), so do we _really_ need to do it each time?

Reviewer 2: Another approach could be to tell the reader to auto-deploy the hub via editing `config.yaml`, and to then confirm the following things work as expected before telling the user representative about it?

the way you would like it to. This might take some iteration and multiple
tries. Specifically test that at least the following work ok:

1. Authentication provider
2. Admin user access

> 3. User image in use

Reviewer: How we can quickly identify the image we are using? In the past, I have "injected" env variables in the Dockerfile to have a quick reference...

> 4. Default user interface when the user logs in (lab, notebook, rstudio, retrolab, etc)

Reviewer: Do we have retrolab? I think it is not actually on our images, but I might be wrong about it...

5. Home page display configuration
6. (Optionally) `dask-gateway` functionality
7. (Optionally) Access to any cloud resources (like storage buckets, etc)

>    granted to the hub users

Reviewer: For the optional ones, we should reference something more specific about the test to perform, I think...

Reviewer 3: Agreed. I always prefer good documentation for stuff I do less frequently as I won't build up the muscle memory like with other tasks. (And it goes without saying, we should have good documentation for everything anyway 😛 )

Reviewer 2: perhaps we could create a new issue to track these? they might be more involved and I think will require separate sections. In particular for dask-gateway and non-standard cloud resources, that will require more expertise and knowledge so we should give those topics extra attention (and this can probably be another PR)

> ## Step 8: Make a PR

Reviewer: Missing blank line before line 62.

Make a PR with your changes, referencing the issue for creation of the hub. Seek
review from someone else, and get this merged!

> Getting this merged will mark the hub as being deployed.

Reviewer 2: ```suggestion
Getting this merged will mark the hub as being deployed.

Finally, notify the users of the hub that it is now deployed, and ask that they test it out to make sure it works as expected.
If they are happy with the result, then close the issue for deploying the new hub.
Congratulations, you are done! 🎉 
```
